---
title: "Alcoholic disease"
author: "[Llu√≠s Revilla](mailto:lrevilla@clinic.cat)"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
  BiocStyle::pdf_document:
    toc: true
---

```{r setup, echo=FALSE, results = "asis", message = FALSE}
knitr::opts_chunk$set(tidy = FALSE, echo = TRUE, cache = TRUE, autodep = TRUE)
```

# Alcohol liver disease
## Load disease expression

The expression data is provided as counts in a tab separate value format. In another file there is the phenoData associated to the samples.

```{r load}
count <- read.delim("../data/ALD_ramon/NGS_AHsteps.expr.counts.tsv", 
                    row.names = 1)
phenoData <- read.csv("../data/ALD_ramon/NGS_AHsteps.design.csv", row.names = 1)
```

First we will explore a little bit the data, in case we need to correct something:

```{r exploration1, message = FALSE}
library("geneplotter")
summary(phenoData)
phenoData
count[1:5, 1:5]
```

So we will delete two colums that has the same information as the column G. And we will use more easily readable names:
```{r pheno}
phenoData <- phenoData[, c("Sample.id", "G")]
colnames(phenoData) <- c("Sample", "Status")
levels(phenoData$Status) <- c("AH", "Non-responders", "Responders", "C.Comp", 
                              "ASH", "Explants", "HCV", "NASH", "Normal")
phenoData
```
However we are only interested on those related to alcohol disease, so we need to subset them:
```{r sub}
alcohol <- !phenoData$Status %in% c("HCV", "NASH", "Explants")
phenoData <- phenoData[alcohol, ]
count <- count[, colnames(count) %in% rownames(phenoData)]
```

We can calculate the normalization factors used to calculate the Counts per million read: 
```{r calcNormFactors}
library("edgeR")
dge <- DGEList(counts = count, group = phenoData$Status)
ord <- order(dge$sample$lib.size)
barplot(dge$sample$lib.size[ord]/1e6, las = 1, ylab = "Millions of reads",
        xlab = "Samples")
dge <- calcNormFactors(dge)
```
We can normalize the data to counts per millon of reads taking into account that we know the normalized library size:
```{r cpm}
expression <- cpm(dge, normalized.lib.sizes = TRUE, log = TRUE, prior.count = 3)
```

## Check quality pre-normalizations
We can observe the quality of the samples with a boxplot and the density of expressions:
```{r q_check}
boxplot(expression, main = "Expression per sample")
multidensity(expression, main = "Densities of expression", legend = NULL)
avgexp <- rowMeans(expression)
hist(avgexp, main = "Histogram of mean expression per gene")
abline(v = 1, col = "red", lwd = 2)
```
There is a lot of differences between samples on the low expressed genes. We filter those genes that are below 1, as they are unreliable:

```{r remove}
expression <- expression[avgexp > 1, ]
multidensity(expression, main = "Densities of expression", legend = NULL)
```
Now the distribution of the expression between samples is much more comparable. 
Following the recomendation of the edgeR package we recalculate the normalization factor without the lowly expressed genes:

```{r recalc}
dge <- calcNormFactors(dge[avgexp > 1, ])
expression <- cpm(dge, normalized.lib.sizes = TRUE, log = TRUE, prior.count = 3)
```

To work easier with the data I proceed to create an ExpressionSet

```{r ESet, message=FALSE}
library("Biobase")
library("AnnotationDbi")
ALD <- ExpressionSet(as.matrix(expression))
phenoData(ALD) <- AnnotatedDataFrame(phenoData)
ALD
```

## Check the normalizations applied 
We take advantatge of the plotMDS function of limma package to plot the MDS of the samples, to see how close are between them. 

```{r plotMDS, message=FALSE }
library("limma")
plotMDS(ALD, labels = phenoData(ALD)$Status, main = "Samples relationships")
```

We can see that they are clearly separated in two groups, but some samples like the Alcohol Hepatitis samples are on both sites of the first dimension. We can also observe that the cirrhosis compensated samples are closer to the normal or healthy samples.  

## Design 

The design we will use to find the differentially expressed genes is the following:
```{r design}
# To ensure sintactically valid names
names <- make.names(levels(droplevels(phenoData(ALD)$Status))) 
design <- sapply(names, function(x){
  x <- ifelse(x == "Non.responders", "Non-responders", x)
  ifelse(phenoData(ALD)$Status == x, 1, 0)
})
rownames(design) <- rownames(phenoData(ALD))
design
```

As you can see each group has its own colum to increase the power to the model, and to allow for more informative contrasts.
```{r design2}
refact <- function(x) {
  if (x == "Normal") {
    0
  } else if ( x == "ASH") {
    1
  } else if (x == "C.Comp") {
    2
  } else if (x == "AH") {
    4
  } else if (x == "Responders") {
    5
  } else if (x == "Non-responders") {
    5
  } else {
    stop("Unexpected level")
  }
}
design2 <- sapply(as.character(phenoData(ALD)$Status), refact)
design2 <- as.matrix(design2)
colnames(design2) <- "Progression"
rownames(design2) <- rownames(phenoData(ALD))
design2
```

Normalize the expression by calculating an appropriate observation weight.
```{r voom}
v <- voom(dge, design, plot = TRUE, normalize.method = "cyclicloess")
```
Note that even if we filter those with low expression we assign the correct library size, to avoid over estimation.

## Estimate surrogate variables
We can further estimate surrogate variables with the package sva:
```{r sva, message = FALSE}
library("sva")
design0 <- model.matrix(~1, data = pData(ALD))
sv <- sva(v$E, design)
sv2 <- sva(v$E, design2)
```
We can append the estimated surrogated variables to the design matrix for a better estimation of the effect of each paramter:

```{r cbind_design}
i.design <- cbind(design, sv$sv)
i.design2 <- cbind(design2, sv$sv)
```

We fit each design with the expression of the alcoholic liver disease:

```{r fitting}
fit <- lmFit(v$E, i.design)
fit.2 <- lmFit(v$E, i.design2)
```

## Contrasts

We are interested in several comparisons, in the first design:
```{r contrasts}
contrasts <- makeContrasts("ASHvsNormal" = ASH - Normal,
                          "CirrhosisVsNormal" = C.Comp - Normal,
                          "AHvsNormal" = AH - Normal,
                          "RespondersVsNormal" = Responders - Normal,
                          "Non-respondersVsNormal" = Non.responders - Normal,
                          
                          # Assuming that some status are similar
                          "SevereVsNormal" = (Responders + Non.responders)/2 - Normal,
                          
                          "RespondersVsNon-responders" = Responders - Non.responders,
                          
                          "CirrhosisVsASH" = C.Comp - ASH,
                          "AHvsCirrhosis" = AH - C.Comp,

                          "Non-respondersVsAH" = Non.responders - AH,
                          "RespondersVsAH" = Responders - AH,
                          "SevereVsHepatitis" = (Non.responders + Responders)/2 - AH,
                          
                          levels = design)
contrasts
```

Given those comparisons of interest we now evaluate the results:

```{r contrast_fit}
fit2 <- contrasts.fit(fit, contrasts)
fit2 <- eBayes(fit2, proportion = 0.1)
results <- decideTests(fit2, adjust.method = "fdr", lfc = log2(2))
summary(results)

fit2.2 <- eBayes(fit.2, proportion = 0.95) 
results2 <- decideTests(fit2.2)
summary(results2)
```

We can observe the quality of the t-values over the teoretical quantiles to observe if there is any assumption about the eBayes fitting which doesn't holds. Note that I already modified the expected proportion of genes in the call to eBayes:
```{r qualtiy, fig.height = 100, fig.width = 30 }
par(mfrow = c(ncol(results), 3))
out <- sapply(colnames(results), function(x){
  qqt(fit2$t[, x], df = fit2$df.total, main = paste("Student's t Q-Q Plot of", x))
  abline(0, 1)
  volcanoplot(fit2, coef = x, main = x)
  plotMD(fit2, coef = x, main = x)
})
```
```{r quality1, fig.width = 20, fig.height = 20 }
par(mfrow = c(ncol(results2), 3))
out <- sapply(colnames(results2), function(x){
  qqt(fit2.2$t[, x], df = fit2.2$df.total, main = paste("Student's t Q-Q Plot of", x))
  abline(0, 1)
  volcanoplot(fit2.2, coef = x, main = paste(x))
  plotMD(fit2.2, coef = x, main = x)
})
```
Here we can see on the last plot that the exprsesion of the genes show a bimodal distribution in the progression of the dissease. Which seems to confirm that at one point it descompensates and has a very bad prognossis. 

## Store data
It is a good idea to store the data, not only the program of your analysis, so here I go:
```{r store_ALD}
save(fit, fit2, fit2.2, fit.2, design, design2, i.design, i.design2, dge, ALD, contrasts, file = "ALD.RData")
```

## Differentially expressed genes

We can plot the top 2000 genes with the highest absolute value of log fold-change in each :
```{r DEG_ALD, fig.width = 30, fig.height = 84 }
par(mfrow = c(ncol(results), 2))
out <- sapply(colnames(results), function(x) {
  tt.ALD <- topTable(fit2, coef = x, sort.by = "logFC", number = Inf)
  tt.ALD <- tt.ALD[order(-abs(tt.ALD$logFC)), ]
  plot(density(tt.ALD[1:2000, "logFC"]), 
       main = paste("Distribution of the top 2000 DEG in", x))
  hist(tt.ALD[1:2000, "logFC"], 
       main = paste("Distribution of the top 2000 DEG in", x))
})
```
For the progression model we can plot it too:
```{r DEG_ALD2, fig.width = 30, fig.height = 30 }
par(mfrow = c(1, 2))
tt.ALD2 <- topTable(fit2.2, coef = "Progression", sort.by = "logFC", number = Inf)
tt.ALD2 <- tt.ALD2[order(-abs(tt.ALD2$logFC)), ]
plot(density(tt.ALD2[1:2000, "logFC"]), main = "Distribution of the top 2000 DEG")
hist(tt.ALD2[1:2000, "logFC"], main = "Distribution of the top 2000 DEG")
```

# SessionInfo

```{r end}
sessionInfo()
```
