---
title: "Ductular Reaction"
author: "[Llu√≠s Revilla](mailto:lrevilla@clinic.cat)"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
  BiocStyle::pdf_document:
    toc: true
---

```{r setup, echo=FALSE, results="asis", message=FALSE}
knitr::opts_chunk$set(tidy = FALSE, echo = TRUE, cache = TRUE, autodep = TRUE)
```

# Introduction

This data was extracted from a liver, the POS labels indicate that were extracted from a microdisection with KRT7 positive. NEG indicates that for the same microdisection those not stained for KRT7 where taken. TOTAL indicates that there is a mix of KRT7+ and KRT7- dyeing.

This is part of a bigger study you can find [here](index.html)

# Load the data 

```{r load}
phenoData <- read.csv("../data/ductular_reaction/PHENO_26_10_2015.csv", 
                      row.names = 1)
expression <- read.delim("../data/ductular_reaction/refseq_counts_DR.csv", 
                         row.names = 1)
```
We explore these data:

```{r exploration, message = FALSE}
library("geneplotter")
summary(phenoData)
phenoData
expression[1:5, 1:5]
```
We can observe the expression uses the RefSeq identifier of NCBI. And that there is a sample identification in the name of  the samples, I proceed to create its own column:

```{r pheno}
phenoData$paired <- as.numeric(substring(rownames(phenoData), 0, 3))
```


We can calculate the normalization factors used to calculate the Counts per million read: 
```{r calcNormFactors}
library("edgeR")
dge <- DGEList(counts = expression, group = phenoData$G)
ord <- order(dge$sample$lib.size)
barplot(dge$sample$lib.size[ord]/1e6, las = 1, ylab = "Millions of reads",
        xlab = "Samples", main = "Library size of the samples", col = phenoData$G)
legend("topleft", legend = levels(phenoData$G), 
       fill = seq_along(levels(phenoData$G)))
dge <- calcNormFactors(dge)
```
We can normalize the data to counts per millon of reads taking into account that we know the normalized library size:
```{r cpm}
logCPM <- cpm(dge, normalized.lib.sizes = TRUE, log = TRUE, prior.count = 3)
```
# Check quality pre-normalizations
We can observe the quality of the samples with a boxplot and the density of expressions:
```{r q_check, fig.width = 30, fig.height = 10}
boxplot(logCPM, main = "Expression per sample", col = phenoData$G)
multidensity(logCPM, main = "Densities of expression", legend = phenoData$G,  
             col = phenoData$G)
avgexp <- rowMeans(logCPM)
hist(avgexp, main = "Histogram of mean expression per gene")
abline(v = 1, col = "red", lwd = 2)
```
There is a lot of differences between samples on the low expressed genes. We filter those genes that are below 1, as they are unreliable:

```{r remove}
logCPM <- logCPM[avgexp > 1, ]
multidensity(logCPM, main = "Densities of expression", legend = NULL)
```
Now the distribution of the expression between samples is much more comparable. 
Following the recomendation of the edgeR package we recalculate the normalization factor without the lowly expressed genes:

```{r recalc, message = FALSE}
dge <- calcNormFactors(dge[avgexp > 1, ])
logCPM <- cpm(dge, normalized.lib.sizes = TRUE, log = TRUE, prior.count = 3)
```
We can check how has this normalization affected the values:
```{r q_check2}
boxplot(logCPM, main = "Expression per sample", col = phenoData$G)
multidensity(logCPM, main = "Densities of expression", legend = NULL)
avgexp <- rowMeans(logCPM)
hist(avgexp, main = "Histogram of mean expression per gene")
```
Now are a little bit more comparable.  
To work easier with the data I proceed to create an ExpressionSet

```{r ESet, message = FALSE}
library("Biobase")
library("AnnotationDbi")
DR <- ExpressionSet(as.matrix(logCPM))
phenoData(DR) <- AnnotatedDataFrame(phenoData)
DR
```

# Check the normalizations applied 
We take advantatge of the plotMDS function of limma package to plot the MDS of the samples, to see how close are between them. 

```{r plotMDS, message=FALSE }
plotMDS(DR, top = Inf, labels = phenoData(DR)$G, main = "Samples relationships")
plotMDS(DR, top = Inf, labels = phenoData(DR)$paired, main = "Samples relationships")
```

We can visualize too, the relationship between samples at gene expression by:
```{r cluster}
colnames(logCPM) <- phenoData(DR)$G
d <- as.dist(1 - cor(logCPM, method = "spearman"))
cluster <- hclust(d)
plot(cluster)

colnames(logCPM) <- phenoData(DR)$paired
d <- as.dist(1 - cor(logCPM, method = "spearman"))
cluster <- hclust(d)
plot(cluster)
colnames(logCPM) <- rownames(phenoData(DR))
```
We can see that some samples tend to be paired by the origin of their data.

# Design
In this case each type of expression has come from a different subset of the liver, so we should consider each one as an independent group. We also add as covariate the patient from where each sample was obtained.
```{r design}
# Grouping by sample origin
design <- model.matrix(~0 + phenoData(DR)$G)
colnames(design) <- c("NEG", "POS", "TOTAL")
# Adding covariate of patient origin
# design <- cbind(design, "paired" = phenoData(DR)$paired) 
design0 <- matrix(1, nrow = nrow(design))
colnames(design0) <- "Intercept"
design
```

And we can adjust the data using the weights previously calculated and the design we just created:
```{r voom}
v <- voom(dge, design, plot = TRUE, normalize.method = "cyclicloess")
```

We can see that the adjusted value:
```{r p-values}
library("sva")
pValues <- f.pvalue(v$E, design, design0)
qValues <- p.adjust(pValues, method = "BH")
hist(qValues, sub = "without using surrogate variables")
```
As we can see, there is a peak in the low p-values for real differentially expressed genes.

# Estimate surrogate variables
We can estimate surrogate variables with the function sva:
```{r sva, message = FALSE}
sv <- sva(v$E, design, design0, method = "irw")
```

We can append the estimated surrogated variables to the design matrix for a better estimation of the effect of each paramter:

```{r cbind_design}
i.design <- cbind(design, sv$sv)
colnames(i.design) <- c(colnames(design), paste0("X", 1:ncol(sv$sv)))
i.design0 <- cbind(design0, sv$sv)
colnames(i.design0) <- c(colnames(design0), paste0("X", 1:ncol(sv$sv)))
```
We can check if now using the estimated surrogate variables makes the null hypothesis more uniform than previously:

```{r p-values-sva}
pValues <- f.pvalue(v$E, i.design, i.design0)
qValues <- p.adjust(pValues, method = "BH")
hist(qValues, sub = "Using surrogate variables")
```
Even after adjusting for surrogate variables it doesn't improve the model. Maybe because we expect big differences between KRT+ and KRT- samples. 
We fit the design with the expression of the ductular reaction, without taking into account the surrogate variables

```{r fitting}
fit <- lmFit(v$E, design)
```

# Contrasts
We are interested in the Positive vs Negative comparison, but we can compute others too:
```{r comp}
contrasts <- makeContrasts("PosVsNeg" = POS - NEG,
                           "PosVsTotal" = POS - TOTAL,
                           "NegVsTotal" = NEG - TOTAL, 
                           levels = design)
contrasts
```

```{r contrast_fit}
fit.2 <- contrasts.fit(fit, contrasts)
fit.2 <- eBayes(fit.2)
results.DR <- decideTests(fit.2, adjust.method = "BH", lfc = log2(2))
summary(results.DR)
```

We can observe the quality of the t values over the teoretical quantiles to observe if there is any assumption about the eBayes fitting which doesn't holds. Note that I already modified the expected proportion of genes in the call to eBayes:
```{r qualtiy, fig.width = 30, fig.height = 40 }
par(mfrow = c(3, 3))
out <- sapply(colnames(results.DR), function(x){
  qqt(fit.2$t[, x], df = fit.2$df.total, main = paste("Student's t Q-Q Plot of", x))
  abline(0, 1)
  volcanoplot(fit.2, coef = x, main = x)
  plotMD(fit.2, coef = x, main = x)
})
```

# Store the data

We store the data in case we need it:
```{r store_DR}
save(fit, fit.2, design, logCPM, DR, contrasts, v, dge, 
     file = "DR.RData")
```

# Differentially expressed genes
```{r DEG_RD, fig.width = 14, fig.height = 30 }
tt.DR <- topTable(fit.2, coef = "PosVsNeg", sort.by = "logFC", number = Inf)
tt.DR <- tt.DR[order(-abs(tt.DR$logFC)), ] # Ordered by the absolute value of logFC
head(tt.DR)
signif <- tt.DR[tt.DR$adj.P.Val < 0.05, ] # Subset of significant p-value
par(mfrow = c(2, 1))
plot(density(signif[, "logFC"]), main = "Distribution of the top 2000 DEG")
hist(signif[, "logFC"], main = "Distribution of the top 2000 DEG")
```

In order to store them we use the symbols, somehow a little bit more informative than the numbers:

```{r RD_tt}
library("org.Hs.eg.db")
symb <- mapIds(org.Hs.eg.db, keys = rownames(tt.DR), keytype = "REFSEQ", column = "SYMBOL")
nam <- mapIds(org.Hs.eg.db, keys = rownames(tt.DR), keytype = "REFSEQ", column = "GENENAME")
DEG <- cbind("Symbol" = symb, "Name" = nam, tt.DR)
write.csv(DEG, file = "DR_PosVsNeg.csv", na = "", row.names = FALSE)
```
We stored all the table with `r nrow(DEG)` genes, even though some are not significant and there might be some sequences that come from the same gene but with a different messenger. 

We can check that KRT7 is more expressed on the positive samples:
```{r KRT7}
DEG[grep("(KRT[0-9])", DEG$Symbol), ]
```

As we would expect, and it is not far from the max log fold-change `r max(signif$logFC)`, in fact is the `r match("KRT7", DEG$Symbol)`th gene more over-expressed.

# Functional enrichment

We can check for functional enrichment, to see if those few significant genes are more related to performing certain functions and processes. I check for pathways using the Reactome data base and the process using Gene Ontologies.

## Reactome
We can see in Reactome which pathways are enriched for those genes, that are significantly differentially expressed.

```{r reactome, collapse=TRUE, fig.width = 20, fig.height = 20}
library("ReactomePA")
library("clusterProfiler")
DEG_up <- mapIds(org.Hs.eg.db, keys = rownames(signif[signif$logFC > 0L, ]), 
                 keytype = "REFSEQ", column = "ENTREZID")
DEG_down <- mapIds(org.Hs.eg.db, keys = rownames(signif[signif$logFC < 0L, ]),
                   keytype = "REFSEQ", column = "ENTREZID")
groups <- list(Up = DEG_up, Down = DEG_down)
cc <- compareCluster(groups, fun = "enrichPathway")
dotplot(cc)
```
We can observe differences in the comparison between genes up-regulated and dow-regulated. 
But for a more general overview is better to see all the genes involved
```{r reactome_all, fig.width = 20, fig.height = 20}
DEG_names <- mapIds(org.Hs.eg.db, keys = rownames(signif), 
                    keytype = "REFSEQ", column = "ENTREZID")
enrich <- enrichPathway(DEG_names, minGSSize = 2, maxGSSize = 2000)
enrich_up <- enrichPathway(DEG_up, minGSSize = 2, maxGSSize = 2000)
enrich_down <- enrichPathway(DEG_down, minGSSize = 2, maxGSSize = 2000)
write.csv(as.data.frame(enrich), file = "Reactome_DR.csv", 
          row.names = FALSE, na = "")
write.csv(as.data.frame(enrich_up), file = "Reactome_up_DR.csv", 
          row.names = FALSE, na = "")
write.csv(as.data.frame(enrich_down), file = "Reactome_down_DR.csv", 
          row.names = FALSE, na = "")
dotplot(enrich)
enrichMap(enrich, layout = igraph::layout_nicely,
          vertex.label.cex = 1, n = 15)
```
Now we can observe which pathways are enriched with those genes, there is a functional difference between the genes up and down regulated. 

## Gene Ontology
We can see the enrichment in the genes ontologies using topGO package:
```{r topGO, eval = TRUE, message = FALSE,  collapse = TRUE, fig.width = 20, fig.height = 20}
library("topGO")
allGenes <- tt.DR[, "adj.P.Val"]
names(allGenes) <- rownames(tt.DR)
topDiffGenes <- function(x) {
  return(x <= 0.05)
}
idsMapping <- mapIds(org.Hs.eg.db, names(allGenes), "GO", keytype = "REFSEQ",
                     multiVals = "list")
GOdata.bp <- new("topGOdata",
                 ontology = "BP",
                 description = "Biological process of the signature module.",
                 allGenes = allGenes,
                 annot = annFUN.gene2GO, gene2GO = idsMapping,
                 geneSelection = topDiffGenes,
                 nodeSize = 5)
save(GOdata.bp, file = "GO_DR.RData")
resultFisher <- runTest(GOdata.bp, algorithm = "classic", statistic = "fisher")
resultKS.weight <- runTest(GOdata.bp, algorithm = 'weight01', statistic = "ks")
resultKS.elim <- runTest(GOdata.bp, algorithm = "elim", statistic = "ks")

allRes <- GenTable(GOdata.bp, classic = resultFisher, 
                   weight01 = resultKS.weight, elim = resultKS.elim, 
                   orderBy = "weight01", topNodes = 1000)
allRes <- allRes[as.numeric(allRes$weight01) < 0.05, ]
write.csv(allRes, file = "GO_DR_diff.csv", row.names = FALSE)

up <- tt.DR[tt.DR$logFC > 0, "adj.P.Val"]
names(up) <- rownames(tt.DR[tt.DR$logFC > 0, ])
GOdata.bp <- new("topGOdata",
                 ontology = "BP",
                 description = "Biological process of the signature module.",
                 allGenes = up,
                 annot = annFUN.gene2GO, gene2GO = idsMapping,
                 geneSelection = topDiffGenes,
                 nodeSize = 5)
save(GOdata.bp, file = "GO_DR_up.RData")
resultFisher <- runTest(GOdata.bp, algorithm = "classic", statistic = "fisher")
resultKS.weight <- runTest(GOdata.bp, algorithm = 'weight01', statistic = "ks")
resultKS.elim <- runTest(GOdata.bp, algorithm = "elim", statistic = "ks")

allRes <- GenTable(GOdata.bp, classic = resultFisher, 
                   weight01 = resultKS.weight,
                   orderBy = "weight01", topNodes = 1000, numChar = 150)
allRes <- allRes[as.numeric(allRes$weight01) < 0.05, ]
write.csv(allRes, file = "GO_DR_up.csv", row.names = FALSE)

down <- tt.DR[tt.DR$logFC < 0, "adj.P.Val"]
names(down) <- rownames(tt.DR[tt.DR$logFC < 0, ])
GOdata.bp <- new("topGOdata",
                 ontology = "BP",
                 description = "Biological process of the signature module.",
                 allGenes = down,
                 annot = annFUN.gene2GO, gene2GO = idsMapping,
                 geneSelection = topDiffGenes,
                 nodeSize = 5)
save(GOdata.bp, file = "GO_DR_down.RData")
resultFisher <- runTest(GOdata.bp, algorithm = "classic", statistic = "fisher")
resultKS.weight <- runTest(GOdata.bp, algorithm = 'weight01', statistic = "ks")
resultKS.elim <- runTest(GOdata.bp, algorithm = "elim", statistic = "ks")

allRes <- GenTable(GOdata.bp, classic = resultFisher, 
                   weight01 = resultKS.weight, elim = resultKS.elim, 
                   orderBy = "weight01", topNodes = 1000, numChar = 200)
allRes <- allRes[as.numeric(allRes$weight01) < 0.05, ]
write.csv(allRes, file = "GO_DR_down.csv", row.names = FALSE)
head(allRes)
showSigOfNodes(GOdata.bp, score(resultKS.weight), firstSigNodes = 10, useInfo = 'all')
title(main = "GO analysis using Weight01 algorithm", line = -2)
```

We can see that there are several GO ontologies related to transport, signaling and regulation of stem cell differenciation, as well as membrane organization. On the plot we can see the relatipn between the top 10 GO calculated using the Weight01 algorithm, some are much closer to the root parent, others are quite specific, as outlied before. 

# SessionInfo

```{r end}
sessionInfo()
```
