---
title: "Weight of ductular reaction"
author: "[Llu√≠s Revilla](mailto:lrevilla@clinic.cat)"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
  BiocStyle::pdf_document:
    toc: true
---

```{r setup, echo=FALSE, results="asis", message=FALSE}
knitr::opts_chunk$set(tidy = FALSE, echo = TRUE, cache = TRUE, autodep = TRUE)
```

# Introduction

To evaluate the weight of the ductular reaction in the steps, we will perform a Gene Set Analysis using the top differentially expressed genes of the ductular reaction on the alcoholic liver disease progression and in the networks build.

To evaluate the relationship of the genes that increase or decrease the expression along the progression of disease with the ductular reaction, we will perform a GSEA on the top differentially expressed genes.

There are several ways to perform gene set enrichment analysis, it may be classified into the following categories:
 - By the null hypothesis they are testing:
   - Competitive: involves the genes inside and outside the gene set.
   - Self-contained: involves only the genes inside the gene set.
 - By whether enrichment score (ES) are calculated using phenotypes or not:
   - Supervised: needs a phenotype to calculate the ES.
   - Unsupervised: doesn't need the phenotype to calculate the ES.
 - By the units that ES are quantifying:
   - The entire sample population: each ES tells something about a gene set and the entire data set.
   - Single sample: each ES tells something about a gene set and an individual sample.
   
Even more, the test can correct for interdependence (by correlation) of the genes in the same gene set or not. 
   
We define a function to perform a simple (competitive, supervised, all samples, uncorrected) gene set enrichment analysis over any statistic, with two methods, one to detect change in the mean expression and another one to detect change in the scales (`method = "category"`):

```{r introduction}
library("GSEABase") # To create gene sets and collections
library("Category") # to use applyByCategory
library("fgsea") # To make GSEAs like in the BROAD webpage
library("org.Hs.eg.db") # To obtain annotation of human sequences
GSEA <- function(gsetcollection, statistic, identifiers = NULL, 
                 method = "simple") {
  if (is.null(names(statistic)) & !is.null(identifiers)) {
    names(statistic) <- identifiers # It should be the same length 
  } else if (is.null(names(statistic)) & is.null(identifiers)) {
    stop("Identifiers are the name of the genes in the statistics, ",
         "please provide them matching with the gset collection")
  }
  identifiers <- names(statistic)
  if (length(statistic) != length(identifiers)) {
    stop("The identifiers don't correspond to the statistic, check the length")
  }
  Im <- incidence(gsetcollection)
  Im <- Im[, colnames(Im) %in% identifiers]
  tGSgenes <- statistic[match(colnames(Im), identifiers)]
  zS <- switch(method, 
               # shifts in mean expression
               simple = sqrt(rowSums(Im)) * (as.vector(Im %*% tGSgenes)/
                                               rowSums(Im)),
               # To detect change in scale
               category = applyByCategory(tGSgenes, Im, function(x) {
                 (sum((x - mean(x))^2) - (length(x) - 1))/(2 * (length(x) - 1))
               })
  )
  zS <- sort(abs(zS), decreasing = TRUE)
  # We assume that the scores follow a normal distribution in 
  # a random group of genes
  pv <- pmin(pnorm(zS), 1 - pnorm(zS)) 
  a.pv <- p.adjust(pv, method = "fdr") # the same as BH
  m <- cbind("zScore" = zS, "p-value" = pv, "FDR" = a.pv)
  m <- m[order(m[, "FDR"], m[, "zScore"], m[, "p-value"],
               decreasing = c(FALSE, TRUE, FALSE)), ]
  as.data.frame(m)
}
```

# Gene Sets of co-expressed genes

## Early stages of ALD

We can load the groups of genes co-expressed in the early stage of the disease:
```{r early_sets}
library("limma")
load("Early_Network/unsigned_signed/modules_ME.RData", verbose = TRUE)
load("Early_Network.RData", verbose = TRUE)
early_modules <- moduleColors
names(early_modules) <- colnames(data.wgcna)

# Make a collection with where each geneSet is a module
gsc_early <- GeneSetCollection(sapply(unique(early_modules), function(x) {
  GeneSet(names(early_modules[early_modules == x]), 
          setName = paste("early", x, sep = "_"),
          geneIdType = SymbolIdentifier("org.Hs.eg.db"))
}))
```

## Late stages of ALD

We can load the groups of genes co-expressed in the late statge of the disease:
```{r late_sets}
load("Late_Network/unsigned_signed/modules_ME.RData", verbose = TRUE)
# load("Late_Network.RData", verbose = TRUE) # Same order of genes!
late_modules <- moduleColors
names(late_modules) <- colnames(data.wgcna)
# Make a collection with each geneSet a module
gsc_late <- GeneSetCollection(sapply(unique(late_modules), function(x) {
  GeneSet(names(late_modules[late_modules == x]), 
          setName = paste("late", x, sep = "_"),
          geneIdType = SymbolIdentifier("org.Hs.eg.db"))
}))
```

# Enrichment in ALD progression

To evaluate the weight of the ductular reaction in the alcoholic liver disease progression we load the [previously](Ductular_reaction.html) calculated differentially expressed genes in the ductular reaction:

```{r load_DR}
DRs <- read.csv("DR_PosVsNeg.csv")
FC <- DRs$logFC
names(FC) <- DRs$Symbol
FC <- FC[!duplicated(names(FC))]
FC <- sort(FC, decreasing = TRUE)
pFC <- FC[DRs$logFC > 0 & DRs$adj.P.Val < 0.05]
nFC <- FC[DRs$logFC < 0 & DRs$adj.P.Val < 0.05]
RD_signature_up <- DRs$Symbol[DRs$logFC > 0 & DRs$adj.P.Val < 0.05][1:100]
RD_signature_dw <- DRs$Symbol[DRs$logFC < 0 & DRs$adj.P.Val < 0.05][1:100]
gst_pFC <- GeneSet(as.character(RD_signature_up), setName = "Top100_Positive_FC",
                   geneIdType = SymbolIdentifier("org.Hs.eg.db"))
gst_nFC <- GeneSet(as.character(RD_signature_dw), setName = "Top100_Negative_FC",
                   geneIdType = SymbolIdentifier("org.Hs.eg.db"))

gs_pFC <- GeneSet(unique(names(pFC)), setName = "Positive_FC",
                  geneIdType = SymbolIdentifier("org.Hs.eg.db"))
gs_nFC <- GeneSet(unique(names(nFC)), setName = "Negative_FC", 
                  geneIdType = SymbolIdentifier("org.Hs.eg.db"))
gsc_FC <- GeneSetCollection(gs_pFC, gs_nFC, gst_nFC, gst_pFC)
```
We can observe where are the genes differentially expressed in the ductular reaction compared to the progressive expression of the genes in the disease:

```{r load_K}
kendall <- read.csv("Kendall_ALD.csv")
K <- kendall$Tau
names(K) <- kendall$Symbol
K.ok <- sort(K, decreasing = TRUE)
K.ok <- K[kendall$p.value < 0.05]
increasing <- K.ok > 0.5
decreasing <- K.ok < -0.5

# We use the file downloaded from broad v5.2 
# to obtain sets from the Kegg, Reactome database
c2BroadSets <- getGmt("../data/c2.all.v5.2.symbols.gmt", SymbolIdentifier("org.Hs.eg.db"))
c2BroadSets <- c2BroadSets[unique(c(grep("^KEGG", names(c2BroadSets)),
                                    grep("^REACTOME", names(c2BroadSets)), 
                                    grep("^BIOCARTA", names(c2BroadSets)),
                                    grep("LIVER", names(c2BroadSets)),
                                    grep("HEPATO", names(c2BroadSets)))
                                  )]
x <- sapply(geneIds(c2BroadSets), length)
c2BroadSets <- c2BroadSets[x >= 20L] # To be able to test it with the GSEA function
c2BroadSets <- mapIdentifiers(c2BroadSets, SymbolIdentifier("org.Hs.eg.db"))
gsc_late <- mapIdentifiers(gsc_late, SymbolIdentifier("org.Hs.eg.db"))
gsc_early <- mapIdentifiers(gsc_early, SymbolIdentifier("org.Hs.eg.db"))
gs <- GeneSetCollection(c(c2BroadSets, gsc_late, gsc_early, gsc_FC))
```


Once we have our gene sets we can calculate the score, we don't need to calculate a scale shift in the tau statistic, so we just compare mean shift:

```{r gsea_K}
meanExpr_K <- GSEA(gs, K)
write.csv(meanExpr_K, "GSEA_meanExpr_K.csv")
head(meanExpr_K)

fgseaRes <- fgsea(pathways = geneIds(gs), stats = K, minSize = 15, 
                 maxSize = Inf, nperm = 10000)
fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
library("data.table") # To be able to store thd ata.table efficiently
fwrite(fgseaRes, file = "fgsea_K.csv")
fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
topPathwaysUp <- fgseaRes[ES > 0 & padj < 0.05][head(order(padj), n = 10),
                                                pathway]
topPathwaysDown <- fgseaRes[ES < 0 & padj < 0.05][head(order(padj), n = 10),
                                                pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
plotGseaTable(geneIds(gs)[topPathways], K, fgseaRes, gseaParam = 0.5)
library("ggplot2")
plotEnrichment(geneIds(gs[[topPathways[1]]]), K) +
  ggtitle(topPathways[1])
plotEnrichment(geneIds(gs[[topPathways[20]]]), K) +
  ggtitle(topPathways[20])
```
We have calculated the enrichment using the scores 

```{r piano}
library("piano")
GeneSetCollection2GSC <- function(gsc, addInfo = "none") {
  GSC <- list("gsc" = geneIds(gsc), "addInfo" = addInfo)
  class(GSC) <- "GSC"
  return(GSC)
}

# Convert every value (pvalue and FC if uses direction) to a matrix!
myPval <- DRs$adj.P.Val
myPval <- as.matrix(myPval, ncol = 1)
rownames(myPval) <- DRs$Symbol
myFC <- DRs$logFC
myFC <- as.matrix(myFC, ncol = 1)
rownames(myFC) <- DRs$Symbol
GS <- GeneSetCollection2GSC(gs)

# Run the competitive, supervised, entire sample
gsaRes <- runGSA(geneLevelStats = myPval, directions = myFC,
                 gsc = GS, nPerm = 1000)
gsaResTab <- GSAsummaryTable(gsaRes)
write.csv(gsaResTab, file = "runGSA_FC.csv")
```

As we can observe the genes over-expressed in the ductular reaction are not enriched in the increase of the expression. We can plot the under-expressed genes and over-expressed genes in the ductular reaction compared to the statistic:

```{r plot_FC_in_K}
index <- ids2indices(geneIds(gs), names(K))

plotEnrichment(geneIds(gs[["Negative_FC"]]), K) +
  ggtitle("Negative FC in DR")

plotEnrichment(geneIds(gs[["Positive_FC"]]), K) +
  ggtitle("Positive FC in DR")
```

# Enrichment in the ductular reaction

We create gene sets of the genes with a progression of the expression in the disease and we use them to build the gene sets. One gene set for the ones that increase and another for the ones which decrease:

```{r K_sets}
gs_iK <- GeneSet(unique(names(K.ok[increasing])), setName = "Increasing_K")
gs_dK <- GeneSet(unique(names(K.ok[decreasing])), setName = "Decreasing_K")
gsc_K <- GeneSetCollection(gs_iK, gs_dK)
gsc_K <- mapIdentifiers(gsc_K, SymbolIdentifier("org.Hs.eg.db"))
save(gs, gsc_K, gsc_early, gsc_late, meanExpr_K, K, DRs, FC, file = "Gene_sets.RData")
```

We can observe the enrichment in the ductular reaction of the previous gene sets. 
Assuming the independence of the genes in each data set:
```{r gsea_FC}
gs <- GeneSetCollection(c(c2BroadSets, gsc_late, gsc_early, gsc_K))
meanExpr <- GSEA(gs, DRs$t, DRs$Symbol)
write.csv(meanExpr, "GSEA_meanExpr_DR_t.csv")
head(meanExpr)
meanExpr <- GSEA(gs, DRs$AveExpr, DRs$Symbol)
write.csv(meanExpr, "GSEA_meanExpr_DR_AveExpr.csv")
head(meanExpr)
meanExpr <- GSEA(gs, FC)
write.csv(meanExpr, "GSEA_meanExpr_DR_logFC.csv")
head(meanExpr)
scaleExpr <- GSEA(gs, DRs$t, DRs$Symbol, method = "category")
write.csv(scaleExpr, "GSEA_scaleExpr_DR_t.csv")
head(scaleExpr)
scaleExpr <- GSEA(gs, DRs$AveExpr, DRs$Symbol, method = "category")
write.csv(scaleExpr, "GSEA_scaleExpr_DR_AveExpr.csv")
head(scaleExpr)
scaleExpr <- GSEA(gs, FC, method = "category")
scaleExpr[scaleExpr$`p-value` < 0.05, ]
write.csv(scaleExpr, "GSEA_scaleExpr_DR_logFC.csv")

fgseaRes <- fgsea(pathways = geneIds(gs), stats = FC, minSize = 15, 
                 maxSize = Inf, nperm = 1000)
fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
fwrite(fgseaRes, file = "fgsea_DR.csv")
topPathwaysUp <- fgseaRes[ES > 0 & padj < 0.05][
  head(order(NES, padj, decreasing = c(TRUE, FALSE)), n = 10), pathway]
topPathwaysDown <- fgseaRes[ES < 0 & padj < 0.05][
  head(order(NES, pval), n = 10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
plotGseaTable(geneIds(gs)[topPathways], FC, fgseaRes, gseaParam = 0.5)
plotEnrichment(geneIds(gs[[topPathways[1]]]), FC) +
  ggtitle(topPathways[1])
plotEnrichment(geneIds(gs[[topPathways[20]]]), FC) +
  ggtitle(topPathways[20])
plotEnrichment(geneIds(gs[["Increasing_K"]]), FC) +
  ggtitle("Increasing K")
plotEnrichment(geneIds(gs[["Decreasing_K"]]), FC) +
  ggtitle("Decreasing K")
```

We can test if there are differences in the way each gene set interacts within each gene set in each group:

```{r gsar, eval=FALSE}
library("GSAR")
# Extract the samples of the contrasts
sel <- design %*% contrasts
keep <- ifelse(sel[, 1] != 0, TRUE, FALSE)
# Subset to the samples of DR + and DR -
pheno <- v$E[, keep]
sel <- sel[keep, ]
# Gene set net correlation analysis, competitive, supervised, entire sample
results <- TestGeneSets(object = pheno,
                        group = ifelse(sel[, 1] == 1L, 1L, 2L),
                        geneSets = geneIds(gs_Refseq), min.size = 20,
                        max.size = Inf, test = "GSNCAtest", nperm = 1000)
gsar.res <- as.data.frame(results)
colnames(gsar.res) <- "pvalue"
write.csv(results, file = "gsar_GSNCAtest.csv")
# Plot the differences between the networks between the samples
# plotMST2.pathway(object = target.pathway, group = c(rep(1,17), rep(2,33)),
#                  name = "late_saddlebrown", group1.name = "MST 1",
#                  group2.name = "MST 2", legend.size = 1, leg.x = -1.2,
#                  leg.y = 2, label.size = 0.8, label.dist = 0.8,
#                  cor.method = "pearson")
```
However it is too slow and we will do it only on specific modules. 

After analysing numerically the relationship between these sets and the statisticals we can plot them:

```{r plot_K_in_FC}
index <- ids2indices(geneIds(gs), names(FC))
barcodeplot(FC, index = index[["Increasing_K"]], 
            index2 = index[["Decreasing_K"]], sub = "(logFC)",
            main = "Progressive genes in the ductular reaction")
barcodeplot(DRs$t, index = index[["Increasing_K"]], 
            index2 = index[["Decreasing_K"]], sub = "(t)",
            main = "Progressive genes in the ductular reaction")
barcodeplot(DRs$AveExpr, index = index[["Increasing_K"]], 
            index2 = index[["Decreasing_K"]], sub = "(AveExpr)",
            main = "Progressive genes in the ductular reaction")
```

Assuming there is some correlation between the genes in each gene set:

```{r gsea_2_FC}
load("DR.RData", verbose = TRUE)
gs_Refseq <- mapIdentifiers(gs, RefseqIdentifier("org.Hs.eg.db"))
index <- ids2indices(geneIds(gs_Refseq), rownames(v$E))
# competitive, supervised, all samples, corrected
c_gs <- camera(v$E, index, design, c(-1, 1, 0), allow.neg.cor = TRUE,
               inter.gene.cor = NA)
write.csv(c_gs, file = "Camera_PosVsNeg.csv", row.names = TRUE)

# competitive, supervised, all samples, uncorrected
r_gs <- romer(v$E, index, design, c(-1, 1, 0))
write.csv(r_gs, file = "romer_in_DR.csv", row.names = TRUE)

g_gst <- sapply(index, geneSetTest, FC)
write.csv(g_gst, file = "geneSetTest_mixed_PosVsNeg.csv")

# self-contained, supervised, all samples, uncorrected
rst_gs <- roast(v$E, index, design, c(-1, 1, 0), set.estatistic = "msq")
write.csv(rst_gs, file = "roast_PosVsNeg.csv")
save(c_gs, r_gs, g_gst, rst_gs, file = "gsea_2_FC.RData")
```
Camera when correcting for the correlation in the gene sets, may set highly co-regulated gene sets that are biological interpretable not at the top of the resulting list. But estimates the direction of the gene set.
Romer uses mean ranks rotation to calculate the statistic. The p-value is estimated by simulation and change each time.

## Gene sets are differentially expressed

In order to know which gene sets are differentially expressed in dhe ductular reaction we ca use a single sample test like GSVA:

```{r gsva_FC}
library("GSVA") # To perform a variation set analysis.
GSexpr <- gsva(exprs(DR), gs_Refseq, rnaseq = TRUE, min.sz = 5, verbose = FALSE)
fit3 <- lmFit(GSexpr$es.obs, design)
fit3.2 <- contrasts.fit(fit3, contrasts)
fit3.2 <- eBayes(fit3.2)
results.DR <- decideTests(fit3.2, adjust.method = "BH")
summary(results.DR)
tt <- topTable(fit3.2, coef = "PosVsNeg", number = Inf)
tt <- tt[order(abs(tt$logFC), decreasing = TRUE), ]
write.csv(tt, file = "setsFC_PosVsNeg.csv", row.names = TRUE)
head(tt)

# Select the new index based on which stage the gene sets where obtained
gsc_e <- GeneSet(names(gsc_early), setName = "Early_network")
gsc_l <- GeneSet(names(gsc_late), setName = "Late_network")
gsc_network <- GeneSetCollection(gsc_e, gsc_l)
index <- ids2indices(geneIds(gsc_network), rownames(tt))

fgseaRes <- fgsea(pathways = geneIds(gsc_network), stats = tt$t, minSize = 15, 
                 maxSize = Inf, nperm = 10000)
fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
fwrite(fgseaRes, file = "fgsea_t_gs.csv")
topPathwaysUp <- fgseaRes[ES > 0 & padj < 0.05][head(order(pval), n = 10),
                                                pathway]
topPathwaysDown <- fgseaRes[ES < 0 & padj < 0.05][head(order(pval), n = 10),
                                                pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
plotGseaTable(geneIds(gs)[topPathways], FC, fgseaRes, gseaParam = 0.5)

meanExpr <- GSEA(gsc_network, tt$t, rownames(tt))
write.csv(meanExpr, "GSEA_meanExpr_t_gs.csv")
head(meanExpr)
meanExpr <- GSEA(gsc_network, tt$logFC, rownames(tt))
write.csv(meanExpr, "GSEA_meanExpr_logFC_gs.csv")
meanExpr <- GSEA(gsc_network, tt$AveExpr, rownames(tt))
write.csv(meanExpr, "GSEA_meanExpr_AveExpr_gs.csv")

scaleExpr <- GSEA(gsc_network, tt$t, rownames(tt), method = "category")
write.csv(scaleExpr, "GSEA_scaleExpr_t_gs.csv")
head(scaleExpr)
scaleExpr <- GSEA(gsc_network, tt$logFC, rownames(tt), method = "category")
write.csv(scaleExpr, "GSEA_scaleExpr_logFC_gs.csv")
scaleExpr <- GSEA(gsc_network, tt$AveExpr, rownames(tt), method = "category")
write.csv(scaleExpr, "GSEA_scaleExpr_AveExpr_gs.csv")

c_gs <- camera(GSexpr$es.obs, index, design, c(-1, 1, 0), 
               allow.neg.cor = TRUE, inter.gene.cor = NA)
write.csv(c_gs, file = "Camera_gs_PosVsNeg.csv", row.names = TRUE, )

r_gs <- romer(GSexpr$es.obs, index, design, c(-1, 1, 0))
write.csv(r_gs, file = "romer_gs_DR.csv", row.names = TRUE, )

rst_gs <- roast(GSexpr$es.obs, index, design, c(-1, 1, 0))
write.csv(rst_gs, file = "roast_gs_PosVsNeg.csv")
```

We can further test if there is an enrichment in the gene sets from the late stages of the disease or from the early stages:

```{r plot_gsva_FC}
barcodeplot(tt$t, index = index[[1]], index2 = index[[2]], 
            main = "Early and Late networks in the DR", sub = "(t)")
barcodeplot(tt$logFC, index = index[[1]], index2 = index[[2]], 
            main = "Early and Late networks in the DR", sub = "(logFC)")
```

## On the modules

We can further test if the modules of the networks are enriched on the top 100 genes of the ductular reaction
```{r fisher}
contingency <- function(gs, signature, size) {
  setIds <- geneIds(gs)
  TT <- sum(setIds %in% signature)
  TF <- length(setIds) - TT
  FT <- length(signature) - TT
  FF <- size - TT - TF - FT
  m <- matrix(c(TT, TF, FT, FF), ncol = 2)
  colnames(m) <- c("GeneSet", "!GeneSet")
  rownames(m) <- c("Signature", "!Signature")
  m
}

tests <- sapply(names(gsc_late), function(x) {
  fisher.test(contingency(gsc_late[[x]], RD_signature_up, 13254))$p.value})
tests <- p.adjust(tests)
tests[tests < 0.05]

tests <- sapply(names(gsc_early), function(x) {
  fisher.test(contingency(gsc_early[[x]], RD_signature_up, 13254))$p.value})
tests <- p.adjust(tests)
tests[tests < 0.05]

tests <- sapply(names(gsc_late), function(x) {
  fisher.test(contingency(gsc_late[[x]], RD_signature_dw, 13254))$p.value})
tests <- p.adjust(tests)
tests[tests < 0.05]

tests <- sapply(names(gsc_early), function(x) {
  fisher.test(contingency(gsc_early[[x]], RD_signature_dw, 13254))$p.value})
tests <- p.adjust(tests)
tests[tests < 0.05]
```

## On each phases

We can observe the top differentially genes of each phase compared with the healthy samples and between each phase of the disease:

```{r phases_DEG}
load("ALD.RData", verbose = TRUE)
tt.ASH <- topTable(fit2, coef = "ASHvsNormal", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.ASH, file = "ASHvsNormal.csv")
tt.Cirr <- topTable(fit2, coef = "CirrhosisVsNormal", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.Cirr, file = "CirrhosisVsNormal.csv")
tt.AH <- topTable(fit2, coef = "AHvsNormal", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.AH, file = "AHvsNormal.csv")
tt.AHsevere <- topTable(fit2, coef = "SevereVsNormal", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.Cirr.ASH, file = "SevereVsNormal.csv")
tt.Cirr.ASH <- topTable(fit2, coef = "CirrhosisVsASH", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.ASH, file = "CirrhosisVsASH.csv")
tt.AH.Cirr <- topTable(fit2, coef = "AHvsCirrhosis", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.AH.Cirr, file = "AHvsCirrhosis.csv")
tt.AHsevere.AH <- topTable(fit2, coef = "SevereVsHepatitis", 
                           sort.by = "logFC", number = Inf)
write.csv(tt.AHsevere.AH, file = "SevereVsHepatitis.csv")
```

We can test with those statistics the enrichment of the gene sets in each comparison:

```{r phases_fgsea, eval = FALSE}
fgsea.plot <- function(fgseaRes) {
  fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
  topPathwaysUp <- fgseaRes[ES > 0 & padj < 0.05][head(order(padj), n = 10),
                                                  pathway]
  topPathwaysDown <- fgseaRes[ES < 0 & padj < 0.05][head(order(padj), n = 10),
                                                    pathway]
  topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
  plotGseaTable(geneIds(gs)[topPathways], K, fgseaRes, gseaParam = 0.5)
}
fgseaRes <- fgsea(pathways = geneIds(gs), stats = tt.ASH$t, minSize = 15, 
                 maxSize = Inf, nperm = 1000)
fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
fwrite(fgseaRes, file = "fgsea_ASH.csv")
fgsea.plot(fgseaRes)
fgseaRes <- fgsea(pathways = geneIds(gs), stats = tt.Cirr$t, minSize = 15, 
                 maxSize = Inf, nperm = 1000)
fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
fwrite(fgseaRes, file = "fgsea_Cirr.csv")
fgsea.plot(fgseaRes)
fgseaRes <- fgsea(pathways = geneIds(gs), stats = tt.AH, minSize = 15, 
                 maxSize = Inf, nperm = 1000)
fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
fwrite(fgseaRes, file = "fgsea_AH.csv")
fgsea.plot(fgseaRes)
fgseaRes <- fgsea(pathways = geneIds(gs), stats = tt.AHsevere, minSize = 15, 
                 maxSize = Inf, nperm = 1000)
fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
fwrite(fgseaRes, file = "fgsea_AHsevere.csv")
fgsea.plot(fgseaRes)
fgseaRes <- fgsea(pathways = geneIds(gs), stats = tt.Cirr.ASH$t, minSize = 15, 
                 maxSize = Inf, nperm = 1000)
fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
fwrite(fgseaRes, file = "fgsea_Cirr.ASH.csv")
fgsea.plot(fgseaRes)
fgseaRes <- fgsea(pathways = geneIds(gs), stats = tt.AH.Cirr$t, minSize = 15, 
                 maxSize = Inf, nperm = 1000)
fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
fwrite(fgseaRes, file = "fgsea_AH.Cirr.csv")
fgsea.plot(fgseaRes)
fgseaRes <- fgsea(pathways = geneIds(gs), stats = tt.AHsevere.AH$t, minSize = 15, 
                 maxSize = Inf, nperm = 1000)
fgseaRes <- fgseaRes[order(abs(NES), padj, decreasing = c(TRUE, FALSE)), ]
fwrite(fgseaRes, file = "fgsea_AHsevere.AH.csv")
fgsea.plot(fgseaRes)
```


We can also observe how are each gene set differentially expressed, including those obtained from the microdissection:
```{r gsva_ALD}
gs <- GeneSetCollection(c(gs, gsc_FC))
GSexpr <- gsva(exprs(ALD), gs, rnaseq = TRUE, min.sz = 5, verbose = FALSE)
fit4 <- lmFit(GSexpr$es.obs, design)
fit4.2 <- contrasts.fit(fit3, contrasts0)
fit4.2 <- eBayes(fit3.2)
```
Once we have transformed the data we can retrieve the fold change for each gene set:
```{r phases_DEGS}
tt.ASH <- topTable(fit4.2, coef = "ASHvsNormal", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.ASH, file = "gs_ASHvsNormal.csv")
tt.Cirr <- topTable(fit4.2, coef = "CirrhosisVsNormal", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.Cirr, file = "gs_CirrhosisVsNormal.csv")
tt.AH <- topTable(fit4.2, coef = "AHvsNormal", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.AH, file = "gs_AHvsNormal.csv")
tt.AHsevere <- topTable(fit4.2, coef = "SevereVsNormal", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.Cirr.ASH, file = "gs_SevereVsNormal.csv")
tt.Cirr.ASH <- topTable(fit4.2, coef = "CirrhosisVsASH", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.ASH, file = "gs_CirrhosisVsASH.csv")
tt.AH.Cirr <- topTable(fit4.2, coef = "AHvsCirrhosis", sort.by = "logFC", 
                    number = Inf)
write.csv(tt.AH.Cirr, file = "gs_AHvsCirrhosis.csv")
tt.AHsevere.AH <- topTable(fit4.2, coef = "SevereVsHepatitis", 
                           sort.by = "logFC", number = Inf)
write.csv(tt.AHsevere.AH, file = "gs_SevereVsHepatitis.csv")
```

# SessionInfo

```{r}
sessionInfo()
```

